pkg/vault/contracts/RouterHooks.sol

```diff
@@ -0,0 +1,402 @@
+    ) external nonReentrant onlyVault returns (uint256 bptAmountOut) {
+        return _initializeHook(params);
+    }
+
+    function _initializeHook(InitializeHookParams calldata params) internal returns (uint256 bptAmountOut) {
+        bptAmountOut = _vault.initialize(
+            params.pool,
+            params.sender,
+            params.tokens,
+            params.exactAmountsIn,
+            params.minBptAmountOut,
+            params.userData
+        );
+
+        for (uint256 i = 0; i < params.tokens.length; ++i) {
+            IERC20 token = params.tokens[i];
+            uint256 amountIn = params.exactAmountsIn[i];
+
+            if (amountIn == 0) {
+                continue;
+            }
+
+            // There can be only one WETH token in the pool.
+            if (params.wethIsEth && address(token) == address(_weth)) {
+                _weth.wrapEthAndSettle(_vault, amountIn);
+            } else {
+                // Transfer tokens from the user to the Vault.
+                // Any value over MAX_UINT128 would revert above in `initialize`, so this SafeCast shouldn't be
+                // necessary. Done out of an abundance of caution.
+                _permit2.transferFrom(params.sender, address(_vault), amountIn.toUint160(), address(token));
+                _vault.settle(token, amountIn);
+            }
+        }
+
+        // Return ETH dust.
+        _returnEth(params.sender);
+    }
+
+    /**
+     * @notice Hook for adding liquidity.
+     * @dev Can only be called by the Vault.
+     * @param params Add liquidity parameters (see IRouter for struct definition)
+     * @return amountsIn Actual amounts in required for the join
+     * @return bptAmountOut BPT amount minted in exchange for the input tokens
+     * @return returnData Arbitrary data with encoded response from the pool
+     */
+    function addLiquidityHook(
+        AddLiquidityHookParams calldata params
+    )
+        external
+        nonReentrant
+        onlyVault
+        returns (uint256[] memory amountsIn, uint256 bptAmountOut, bytes memory returnData)
+    {
+        return _addLiquidityHook(params);
+    }
+
+    function _addLiquidityHook(
+        AddLiquidityHookParams calldata params
+    ) internal returns (uint256[] memory amountsIn, uint256 bptAmountOut, bytes memory returnData) {
+        (amountsIn, bptAmountOut, returnData) = _vault.addLiquidity(
+            AddLiquidityParams({
+                pool: params.pool,
+                to: params.sender,
+                maxAmountsIn: params.maxAmountsIn,
+                minBptAmountOut: params.minBptAmountOut,
+                kind: params.kind,
+                userData: params.userData
+            })
+        );
+
+        // maxAmountsIn length is checked against tokens length at the Vault.
+        IERC20[] memory tokens = _vault.getPoolTokens(params.pool);
+
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            IERC20 token = tokens[i];
```

Comment:
Same as above; could move assignment below "continue" (like it's done in remove liquidity)

Reply:
(same, let's not modify the code in this PR)